src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,153 @@
+        }
+
+        public void DisposeLock(IDistributedLock @lock) {
+            var childLifetimeScope = CreateChildLifetimeScope(@lock.Name);
+
+            try {
+                var repository = childLifetimeScope.Resolve<IRepository<LockRecord>>();
+                var transactionManager = childLifetimeScope.Resolve<ITransactionManager>();
+                transactionManager.RequireNew(IsolationLevel.ReadCommitted);
+                var record = repository.Get(@lock.Id);
+
+                if (record != null) {
+                    if (record.ReferenceCount > 0)
+                        record.ReferenceCount--;
+                }
+            }
+            catch (Exception ex) {
+                if (ex.IsFatal()) throw;
+                Logger.Error(ex, "An non-fatal error occurred while trying to dispose a distributed lock with name {0} and ID {1}.", @lock.Name, @lock.Id);
+            }
+            finally {
+                childLifetimeScope.Dispose();
+            }
+        }
+
+        protected virtual bool TryAcquireLockRecord(string name, TimeSpan maxLifetime, out IDistributedLock @lock) {
+            @lock = null;
+            var childLifetimeScope = CreateChildLifetimeScope(name);
+
+            try {
+                var transactionManager = childLifetimeScope.Resolve<ITransactionManager>();
+                transactionManager.RequireNew(IsolationLevel.ReadCommitted);
+
+                // This way we can create a nested transaction scope instead of having the unwanted effect
+                // of manipulating the transaction of the caller.
+                var repository = childLifetimeScope.Resolve<IRepository<LockRecord>>();
+
+                // Find an existing, active lock, if any.
+                var record = repository.Table.FirstOrDefault(x => x.Name == name && (x.ValidUntilUtc >= _clock.UtcNow || x.ReferenceCount > 0));
+
+                // The current owner name (based on machine name and current thread ID).
+                var currentOwnerName = GetOwnerName();
+                var canAcquireLock = false;
+
+                // Check if there's already an active lock.
+                if (record != null) {
+                    // Check if the owner of the lock is the one trying to acquire the lock.
+                    if (record.Owner == currentOwnerName) {
+                        record.ReferenceCount++;
+                        canAcquireLock = true;
+                    }
+                }
+                else {
+                    // No one has an active lock yet, so good to go.
+                    record = new LockRecord {
+                        Name = name,
+                        Owner = currentOwnerName,
+                        ReferenceCount = 1,
+                        CreatedUtc = _clock.UtcNow,
+                        ValidUntilUtc = _clock.UtcNow + maxLifetime
+                    };
+                    repository.Create(record);
+                    repository.Flush();
```

Comment:
Yeah we actually don't need to Flush anymore since the lifetime scope is being disposed of in this very method anyway.